sway/input/seatop_default.c

```diff
@@ -200,25 +200,26 @@ static void cursor_do_rebase(struct sway_cursor *cursor, uint32_t time_msec,
 		if (seat_is_input_allowed(cursor->seat, surface)) {
 			wlr_seat_pointer_notify_enter(wlr_seat, surface, sx, sy);
 		}
-	} else if (node && node->type == N_CONTAINER) {
-		// Try a node's resize edge
-		enum wlr_edges edge = find_resize_edge(node->sway_container, surface, cursor);
-		if (edge == WLR_EDGE_NONE) {
-			cursor_set_image(cursor, "left_ptr", NULL);
-		} else if (container_is_floating(node->sway_container)) {
-			cursor_set_image(cursor, wlr_xcursor_get_resize_name(edge), NULL);
-		} else {
-			if (edge & (WLR_EDGE_LEFT | WLR_EDGE_RIGHT)) {
-				cursor_set_image(cursor, "col-resize", NULL);
+	} else {
+		if (node && node->type == N_CONTAINER) {
```

Comment:
Thanks for you PR!

Why is this block changed? It would make git-bisect's life easier if only one line was modified to add a call to `wlr_seat_pointer_notify_enter`.

Reply:
I guess you mean `git-blame` here, right?

My motivation for this was readability, but TBH `cursor_do_rebase()` doesn't seem to be the correct place to notify enter/leave events. So, as an improvement I'd propose to either (1) move event notification out of this function to where it's called from, or (2) rename this function to better reflect what it actually does. wdyt?
Anyways, if you still prefer just to add the missing call, I'll proceed and do it :)

Thanks

Reply:
>I guess you mean git-blame here, right?

Yeah, git-blame too. When searching for the cause of a bug, it'll be harder if a commit contains unrelated refactoring.

>move event notification out of this function to where it's called from

Interestingly, @Xyene suggested the same thing.

"Rebasing" here means that "update the pointer to be consistent with whatever is under it". It's used when changing the layout for instance, or after a TTY switch. If we remove enter/leave events from this function, we could rename it to something like `cursor_update_image` and re-use it from the tablet code.

Reply:
> Yeah, git-blame too. When searching for the cause of a bug, it'll be harder if a commit contains unrelated refactoring.

Ack and done.

> If we remove enter/leave events from this function, we could rename it to something like cursor_update_image and re-use it from the tablet code.

Sounds like something worth doing. Will send it in a followup patch soon.
Thanks

Reply:
:+1: 

To summarize a discussion from IRC yesterday (since I imagine you might find it interesting):

* the cursor/seat code predates tablet and touch specs, so is tightly coupled with pointers
* this is "surprising", e.g. you'd expect functions like `cursor_motion` or `cursor_do_rebase` to not send pointer events, but they do
   * this means they're not used for tablet/touch support, instead both tablet and touch re-implement ad-hoc parts of that functionality
   * since fewer people use those input methods, it's easier for the code to fall behind without anyone noticing
* tablet support, at least, currently has [some issues](https://github.com/swaywm/sway/issues?q=is%3Aopen+is%3Aissue+author%3AXyene+tablet)<sup>1</sup> that would likely best be fixed by decoupling the `*cursor*` functions from sending pointer events, and moving those events to the call sites

The plan proposed looks something like this:

1. move the pointer events out of the cursor functions
2. reuse them for tablet support &mdash; this'd fix a bunch of tablet issues
3. for future maintainability, add handlers for tablet and touch into seatop, so that the logic for those devices can be migrated from `cursor.c`
4. once everything's in one place rather than two and confirmed "not broken by refactoring", it'll be easier to see what abstractions make sense

<sup>1</sup> There are more if you just search for "tablet", but these are the ones I know the cause of so far :wink:

Reply:
Hi @Xyene, thanks for the info. It sounds like a good plan for me.